### PR TITLE
Shooter class with mock subsystems

### PR DESCRIPTION
**Pull request by Barrelwolf38087**
Basically, there are a bunch of flags that indicate the state of the shooter. All of the public methods except `periodic()` simply set the flags and execute very quickly. `periodic()` is called in the robot loop, and contains all of the logic for what we need to do with those flags. It should also execute quickly. It uses WPILib's Timer class for delays. It supports [input buffering](https://www.ssbwiki.com/Buffer), as well as cancelling buffered actions.

### CODE REVIEW

File: src/main/java/frc/robot/subsystems/Shooter.java [hunk 1]
```diff
@@ -0,0 +1,218 @@
...
+        m_spinUpTimer = new Timer();
+    }
+
+    /**
+     * Call this in the robot loop.
+     * TODO: Consider replacing this home-grown timing system with WPILib's scheduler.
+     */
+    public void periodic() {
+
+        // Check if we're done spinning up yet
```
Comment (jhinspeter) [inline src/main/java/frc/robot/subsystems/Shooter.java:106]: Let's move to an enum-based state machine implemented with a switch statement

File: src/main/java/frc/robot/subsystems/Shooter.java [hunk 2]
```diff
@@ -0,0 +1,218 @@
...
+        void set(TurretPosition position);
+    }
+
+    @FunctionalInterface
+    private interface Vision {
+        TurretPosition calcTargetPosition();
+    }
+
+    private interface Storage {
+        boolean isBallStopOpen();
```
Comment (jhinspeter) [inline src/main/java/frc/robot/subsystems/Shooter.java]: Probably not needed

File: src/main/java/frc/robot/subsystems/Shooter.java [hunk 3]
```diff
@@ -0,0 +1,218 @@
...
+            if (m_ready) {
+                // Log a warning if we get into a weird state. This should never happen.
+                // TODO: Use Max's logging system
+                if (m_spinningUp)
+                    System.err.println("WARNING: Shooter: Both the ready and spinning up flags are set! This shouldn't be possible!");
+
+                m_firing = true;
+                m_fireTimer.start();
+                m_storage.releaseBallStop();
+                --m_buffer;
```
Comment (jhinspeter) [inline src/main/java/frc/robot/subsystems/Shooter.java]: Let's change to variable++, variable--

**Review by jhinspeter — DISMISSED** [review]

**Comment by joecaraccio** [comment]
I don't know if our hardware is going to be as 1:1 as we'd hope. meaning we might not be able to shoot one ball at a time. We might also want to think about the ability to ramp voltage..

Its likely we are going to loose power on our wheel as balls go through quickly, we might want some sort of functionality to ramp voltage over a few seconds

**Comment by Barrelwolf38087** [comment]
That's not too far from what it already does. It's currently configured to wait half a second between firing power cells. The duration it waits is also configurable, so I don't think firing one at a time will be that difficult. We just need to tune the value.

**Comment by joecaraccio** [comment]
I don't know yet what the hardware will look like, but from what Brian has says, we are going to basically have a stopper between the elevator and the shooter. I don't think we can have a ball in the shooter and then spin the wheel (think about a Frisbee shooter with the frisbee contacting the wheel before its even running).  

From the sounds of it we won't have the control to shoot one at a time, if that makes sense.. once we open the chamber it would be hard to know when to push our numatic piston back in to stop the flow of balls (without somehow spearing a ball in the chamber or something).

File: src/main/java/frc/robot/subsystems/Shooter.java [hunk 4]
```diff
@@ -0,0 +1,218 @@
...
+
+    // State variables
+    private boolean m_ready = false;
+    private boolean m_spinningUp = false;
+    private boolean m_firing = false;
+    private int m_buffer = 0;
+    private Timer m_spinUpTimer;
+    private Timer m_fireTimer;
+
+    public static Shooter getInstance() {
```
Comment (ajwharton) [inline src/main/java/frc/robot/subsystems/Shooter.java]: synchronized

File: src/main/java/frc/robot/subsystems/Shooter.java [hunk 5]
```diff
@@ -0,0 +1,218 @@
...
+            + ", "
+            + position.getElevation()
+            + ")"
+        );
+        m_vision = () -> {
+            System.out.println("Getting dummy vision target");
+            return new TurretPosition(0, 0);
+        };
+        m_storage = new Storage() {
+            private boolean open = false;
```
Comment (ajwharton) [inline src/main/java/frc/robot/subsystems/Shooter.java]: m_open?

File: src/main/java/frc/robot/subsystems/Shooter.java [hunk 6]
```diff
@@ -0,0 +1,218 @@
...
+
+            @SuppressWarnings("FieldCanBeLocal")
+            private boolean full = false;
```
Comment (ajwharton) [inline src/main/java/frc/robot/subsystems/Shooter.java]: m_isFull?

File: src/main/java/frc/robot/subsystems/Shooter.java (hunk 4 above)
Comment (Barrelwolf38087) [reply src/main/java/frc/robot/subsystems/Shooter.java]: done

File: src/main/java/frc/robot/subsystems/Shooter.java (hunk 5 above)
Comment (Barrelwolf38087) [reply src/main/java/frc/robot/subsystems/Shooter.java]: got rid of this one, wasn't used anywhere

File: src/main/java/frc/robot/subsystems/Shooter.java (hunk 6 above)
Comment (Barrelwolf38087) [reply src/main/java/frc/robot/subsystems/Shooter.java]: also got rid of this one

File: src/main/java/frc/robot/subsystems/Shooter.java (hunk 2 above)
Comment (Barrelwolf38087) [reply src/main/java/frc/robot/subsystems/Shooter.java]: removed

File: src/main/java/frc/robot/subsystems/Shooter.java (hunk 3 above)
Comment (Barrelwolf38087) [reply src/main/java/frc/robot/subsystems/Shooter.java]: done

**Comment by Barrelwolf38087** [comment]
@jhinspeter should I merge it now?

**Review by jhinspeter — APPROVED** [review]

**Pull request merged**